File: its-my-training-day/leetcode-top-75/questions/intervals/intervals-merge/insert-interval/solution.ts
import {
  ArrayList,
  Stack,
  Queue,
  ListNode,
  TreeNode,
  MinHeap,
  MaxHeap,
} from "../../../../shared/typescript/ds";

const quickSortIntervals = (intervals: number[][], left: number, right: number): void => {
  if (left >= right) return;
  const pivot = intervals[Math.floor((left + right) / 2)][0];
  let i = left;
  let j = right;
  while (i <= j) {
    while (intervals[i][0] < pivot) i += 1;
    while (intervals[j][0] > pivot) j -= 1;
    if (i <= j) {
      const temp = intervals[i];
      intervals[i] = intervals[j];
      intervals[j] = temp;
      i += 1;
      j -= 1;
    }
  }
  if (left < j) quickSortIntervals(intervals, left, j);
  if (i < right) quickSortIntervals(intervals, i, right);
};
export const insertInterval = (intervals: number[][], newInterval: number[]): number[][] => {
  const result = new ArrayList<number[]>();
  let i = 0;
  while (i < intervals.length && intervals[i][1] < newInterval[0]) {
    result.add(intervals[i]);
    i += 1;
  }
  while (i < intervals.length && intervals[i][0] <= newInterval[1]) {
    if (intervals[i][0] < newInterval[0]) newInterval[0] = intervals[i][0];
    if (intervals[i][1] > newInterval[1]) newInterval[1] = intervals[i][1];
    i += 1;
  }
  result.add(newInterval);
  while (i < intervals.length) {
    result.add(intervals[i]);
    i += 1;
  }
  return toArray(result);
};

File: its-my-training-day/leetcode-top-75/questions/intervals/intervals-merge/insert-interval/solution.py
import os
import sys
from typing import List, Dict, Optional, Tuple

ROOT_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../../"))
if ROOT_DIR not in sys.path:
    sys.path.append(ROOT_DIR)

from shared.python.ds import ArrayList, ListNode, Stack, Queue, TreeNode, MinHeap, MaxHeap

def insert_interval(intervals: List[List[int]], new_interval: List[int]) -> List[List[int]]:
    result = ArrayList()
    i = 0
    while i < len(intervals) and intervals[i][1] < new_interval[0]:
        result.add(intervals[i])
        i += 1
    while i < len(intervals) and intervals[i][0] <= new_interval[1]:
        if intervals[i][0] < new_interval[0]:
            new_interval[0] = intervals[i][0]
        if intervals[i][1] > new_interval[1]:
            new_interval[1] = intervals[i][1]
        i += 1
    result.add(new_interval)
    while i < len(intervals):
        result.add(intervals[i])
        i += 1
    return result.to_list()
